### PR TITLE
refactor(portfolio): balance formatting

### PR DESCRIPTION
**Pull request by zsh28**
----

> [!IMPORTANT]
> Refactor balance formatting using `Intl.NumberFormat` for precision and abbreviation, add tests, and update dependencies for Vitest.
> 
>   - **Behavior**:
>     - Refactor `formatBalance` and `formatBalanceUsd` in `format-balance.ts` and `format-balance-usd.ts` to use `Intl.NumberFormat` for better precision and large number abbreviations (M, B, T).
>     - Handle very small amounts with thresholds (e.g., `<0.00001`, `<$0.01`).
>     - Consistent precision control up to 6 decimals for tokens.
>   - **Testing**:
>     - Add `format-balance.test.ts` to test `formatBalance`, `formatBalanceUsd`, and `formatUsdValue` functions for various cases including edge cases and negative values.
>   - **Dependencies**:
>     - Add `vitest` and `@workspace/config-vitest` to `package.json` for testing.
>     - Add `vitest.config.mts` for Vitest configuration.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=samui-build%2Fsamui-wallet&utm_source=github&utm_medium=referral)<sup> for 7e585d6b6d52f622fb5548a6bfe8be5196ede461. You can [customize](https://app.ellipsis.dev/samui-build/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by tobeycodes** [comment]
@zsh28 Thank you for opening this pull request. We will review this once we have submitted the project to the hackathon.

**Review by tobeycodes — CHANGES_REQUESTED** [review]
@zsh28 This looks like it needs some changes. I am seeing duplicate `$` signs in some cases; numbers are being stripped. 

Before

image

After

image

Example of how solscan will shows these

image

It might be a good idea to take a look at phantom, backpack and solflare to see how these are displayed.

If we are going to maintain this code ourselves please make sure to include enough test coverage for a wide range of number variations. If it makes more sense to use cryptoformat and they meet our needs, then that is fine with me. You can also spend some time researching other packages around as there are probably others who have also solved this too.

**Review by tobeycodes — CHANGES_REQUESTED** [review]
Tests are failing

File: packages/portfolio/src/data-access/format-balance.ts [hunk 1]
```diff
@@ -1,6 +1,158 @@
...
+const formatters = new Map<string, Intl.NumberFormat>()
+
+export function formatBalance({
+  balance = 0,
+  decimals,
+}: {
+  balance: bigint | number | undefined
+  decimals: number
+}): string {
+  const balanceNum = typeof balance === 'bigint' ? balance : BigInt(balance || 0)
```
Comment (tobeycodes) [inline packages/portfolio/src/data-access/format-balance.ts]: **request:** this is complex. let's only accept one type as input for this function to reduce complexity

this should be applied throughout the PR

File: packages/portfolio/src/data-access/format-balance.ts [hunk 2]
```diff
@@ -1,6 +1,158 @@
-export function formatBalance({ balance = 0, decimals }: { balance: bigint | number | undefined; decimals: number }) {
-  const balanceString = balance.toString()
-  const integerPart = balanceString.length > decimals ? balanceString.slice(0, balanceString.length - decimals) : '0'
-  const fractionalPart = balanceString.padStart(decimals, '0').slice(-decimals)
-  return `${integerPart}.${fractionalPart}`
+const formatters = new Map<string, Intl.NumberFormat>()
```
Comment (tobeycodes) [inline packages/portfolio/src/data-access/format-balance.ts]: **question:** why are these formatter maps needed?

File: packages/portfolio/src/data-access/format-balance.ts [hunk 3]
```diff
@@ -1,6 +1,158 @@
...
+  }
+
+  return getUsdFormatter({
+    maximumFractionDigits: 7,
+    minimumFractionDigits: 0,
+  }).format(usdValue)
+}
+
+export function formatUsdValue(usdValue: number): string {
+  // Handle special cases
```
Comment (tobeycodes) [inline packages/portfolio/src/data-access/format-balance.ts]: **request:** make this comment clear. what are the special cases and why are they needed?

**Review by tobeycodes — CHANGES_REQUESTED** [review]
This still does not appear to be working how I would expect:
- SOL balance of 0.695954101 is shown with only 4 decimal places
- USD values are shown with more than 2 decimal places.

Image

Image

I could be wrong but I feel like the Cyclomatic complexity and N-path complexity is quite high here. I would really love to see if we can land on a simpler solution here. I could be wrong but I feel like a lot of this code was generated by an LLM without enough context

Please also resolve issues from Github copilot prior to requesting review from me or @beeman.

**Comment by zsh28** [comment]
@tobeycodes apologies for the delay solved the issues had to revert all the changes and re add it back it should be good now please check and let me know

File: packages/portfolio/src/data-access/format-balance-usd.ts [hunk 4]
```diff
@@ -1,14 +1,77 @@
...
     maximumFractionDigits: 2,
-    minimumFractionDigits: 0,
-  }).format(((usdPrice ?? 0) * Number(balance)) / 10 ** decimals)
+    minimumFractionDigits: 2,
+  }).format(absValue)
+
+  return isNegative ? `-${formatted}` : formatted
+}
+
+function getUsdFormatter(options: Intl.NumberFormatOptions): Intl.NumberFormat {
```
Comment (beeman) [inline packages/portfolio/src/data-access/format-balance-usd.ts]: Export to a separate file `get-usd-formatter.ts`

File: packages/portfolio/src/data-access/format-balance.ts [hunk 5]
```diff
@@ -1,6 +1,53 @@
...
+  }
+
+  // Standard token formatting with up to 5 decimals
+  return getTokenFormatter({
+    maximumFractionDigits: 5,
+    minimumFractionDigits: 0,
+  }).format(value)
+}
+
+function getTokenFormatter(options: Intl.NumberFormatOptions): Intl.NumberFormat {
```
Comment (beeman) [inline packages/portfolio/src/data-access/format-balance.ts]: Export to a separate file `get-token-formatter.ts`

**Review by beeman — COMMENTED** [review]
Hey @zsh28 - left some small comments on the PR

File: packages/portfolio/src/data-access/format-balance-usd.ts [hunk 6]
```diff
@@ -1,14 +1,38 @@
...
+
+  // Handle undefined explicitly to avoid unnecessary BigInt creation or precision loss
+  let tokenValue: number
+  if (typeof balance === 'bigint') {
+    if (balance === 0n) {
+      return '$0.00'
+    }
+    // Use string manipulation for large bigints to avoid precision loss
+    tokenValue = balanceToNumber(balance, decimals)
+  } else if (typeof balance === 'number') {
```
Comment (beeman) [inline packages/portfolio/src/data-access/format-balance-usd.ts]: Avoid using `let` and `else` statements.

Instead of saving `tokenValue` in a temp variable and re-assign it, just return the `formatUsdValue` with the new token value for that case.

This will allow us to get rid of all the `else` statements and the `let`, which makes the code shorter, flatter (less indents) and way easier to grasp.

File: packages/portfolio/src/data-access/get-usd-formatter.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+const usdFormatters = new Map<string, Intl.NumberFormat>()
+
+export function getUsdFormatter(options: Intl.NumberFormatOptions): Intl.NumberFormat {
+  // Create a unique key for the formatter based on options
```
Comment (beeman) [inline packages/portfolio/src/data-access/get-usd-formatter.ts]: Unnecessary comment (looks like added by an AI)

File: packages/portfolio/src/data-access/format-usd-value.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { getUsdFormatter } from './get-usd-formatter.ts'
+
+export function formatUsdValue(usdValue: number): string {
+  if (!Number.isFinite(usdValue) || usdValue === 0) {
+    return '$0.00'
+  }
+
+  // Handle negative values by working with absolute amount then applying sign
```
Comment (beeman) [inline packages/portfolio/src/data-access/format-usd-value.ts]: Unnecessary comment (looks like added by an AI)

**Review by beeman — CHANGES_REQUESTED** [review]
Few more improvements and then we can land this.

File: packages/portfolio/src/data-access/get-token-formatter.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+const tokenFormatters = new Map<string, Intl.NumberFormat>()
+
+export function getTokenFormatter(options: Intl.NumberFormatOptions): Intl.NumberFormat {
+  // Build efficient cache key from the options we actually use
```
[the remark this answers is not in the document]
Comment (zsh28) [reply packages/portfolio/src/data-access/get-token-formatter.ts]: removed comment

File: packages/portfolio/src/data-access/format-usd-value.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  const isNegative = usdValue < 0
```
[the remark this answers is not in the document]
Comment (zsh28) [reply packages/portfolio/src/data-access/format-usd-value.ts]: yes I remved that `isNegative` variable in this file

File: packages/portfolio/src/data-access/format-balance-usd.ts (hunk 6 above)
Comment (zsh28) [reply packages/portfolio/src/data-access/format-balance-usd.ts]: yes I got rid of this and its a lot more simplified

File: packages/portfolio/src/data-access/format-balance-usd.ts [hunk 11]
```diff
@@ -1,14 +1,38 @@
...
   balance: bigint | number | undefined
   decimals: number
-  usdPrice: number
+  usdPrice: number | undefined
 }) {
-  return new Intl.NumberFormat('en-US', {
-    maximumFractionDigits: 2,
-    minimumFractionDigits: 0,
-  }).format(((usdPrice ?? 0) * Number(balance)) / 10 ** decimals)
+  if (!usdPrice) {
```
[the remark this answers is not in the document]
Comment (zsh28) [reply packages/portfolio/src/data-access/format-balance-usd.ts]: solved this and added the conditions like this as you have suggested

File: packages/portfolio/src/data-access/format-balance-usd.ts (hunk 4 above)
Comment (zsh28) [reply packages/portfolio/src/data-access/format-balance-usd.ts]: done

File: packages/portfolio/src/data-access/utils.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+export function balanceToNumber(balance: bigint, decimals: number): number {
+  // Convert bigint to string for manipulation
+  const balanceStr = balance.toString()
+  const isNegative = balance < 0n
+  const absStr = isNegative ? balanceStr.slice(1) : balanceStr
```
[the remark this answers is not in the document]
Comment (zsh28) [reply packages/portfolio/src/data-access/utils.ts]: changed naming to `absoluteString`

File: packages/portfolio/src/data-access/format-balance-usd.ts [hunk 13]
```diff
@@ -1,14 +1,87 @@
...
+
+  // Handle undefined explicitly to avoid unnecessary BigInt creation
+  const balanceNum = typeof balance === 'bigint' ? balance : balance ? BigInt(balance) : 0n
+
+  if (balanceNum === 0n) {
+    return '$0.00'
+  }
+
+  // Use string manipulation for large bigints to avoid precision loss
+  const tokenValue = balanceToNumber(balanceNum, decimals)
```
[the remark this answers is not in the document]
Comment (zsh28) [reply packages/portfolio/src/data-access/format-balance-usd.ts]: alot has changed from the suggestions from Beeman this can be ignored

File: packages/portfolio/src/data-access/format-balance.ts [hunk 14]
```diff
@@ -1,6 +1,158 @@
...
+}
+
+export function formatUsdValue(usdValue: number): string {
+  // Handle special cases
+  if (Number.isNaN(usdValue) || !Number.isFinite(usdValue)) {
+    return '$0'
+  }
+
+  // Handle zero case
+  if (usdValue === 0) return '$0'
```
[the remark this answers is not in the document]
Comment (zsh28) [reply packages/portfolio/src/data-access/format-balance.ts]: sure

**Review by beeman — APPROVED** [review]
